### ISSUE TITLE
Support for plain text and json parts in a multipart-form upload

### ISSUE THREAD

**Issue by chrisco314** [Alamofire/Alamofire#591]
### Support for plain text and json parts in a multipart-form upload

**TL;DR:**

I needed support for text / json parts in a multipart-form upload, added a single function to do so, and forked Alamofire here: https://github.com/chrisco314/Alamofire/tree/multipart_data_form_support_for_additional_data_types to do so.

**Context:**

I posted a question to SO here: http://stackoverflow.com/questions/31442411/alamofire-support-for-uploading-multipart-forms-with-text-and-file-parts in my efforts to figure out how to upload textual data in tandem with binary data, but got no response.

**Implementation:**

In the end, to achieve the compact form of:

    form.appendBodyPart(Multipart.plain("patient", text:"\(last.id!)")!)

I needed to add a single override to MultipartFormData.swift:

    /**
    Creates a body part from the data and appends it to the multipart form data object.

    The body part data will be encoded using the following format:

    - `Content-Disposition: form-data; name=#{name}` (HTTP Header)
    - Encoded file data
    - Multipart form boundary

    :param: headers The headers to encode into the `multipart-form` part.
    :param: data The data to encode into the `multipart-form` part.
    */
    public func appendBodyPart(#headers:[String: String], data: NSData) {
        
        let bodyStream = NSInputStream(data: data)
        let bodyContentLength = UInt64(data.length)
        let bodyPart = BodyPart(headers: headers, bodyStream: bodyStream, bodyContentLength: bodyContentLength)
        self.bodyParts.append(bodyPart)
    }

This allowed me to add support for text and json data payloads through the code pasted below. The approach I took was to build up abstracted multipart-data parts through the use of typesafe enums and method chaining:

        class MultipartPlain : MultipartProtocol {
            
            init?(name:String, text:String) {
                
                // initialize header
                self.headers = MultipartHeader()
                    .contentType(.Plain)
                    .contentDisposition(.Name(name))
                
                // serialize body
                self.body = text.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)
                if self.body == nil {
                    return nil
                }
            }
            
            /// The headers used to generate a `text/plain` part in a `multipart/form-data` part
            let headers:MultipartHeader
            
            /// The body used to generate a `text/plain` part in a `multipart/form-data` part
            var body:NSData!
        }

The extension code in its entirety is:

    import Alamofire

    /**
    Extension to the Alamofire MultipartFormData type to support direct
    construction of `multipart/form-data` parts
    */
    extension MultipartFormData {
        
        func appendBodyPart(part:MultipartProtocol) {
            
            let headers = part.headers
            let bodyStream = NSInputStream(data: part.body)
            let bodyContentLength = UInt64(part.body.length)
            appendBodyPart(headers:headers.headers, data:part.body)
        }
    }

    /**
    Protocol that defines the behavior of `multipart/form-data` parts, allowing
    extension to parts of different types, while hiding implementation details
    */
    protocol MultipartProtocol {
        /// The headers used to generate a part in a `multipart/form-data` part
        var headers:MultipartHeader { get }
        
        /// The body used to generate a part in a `multipart/form-data` part
        var body:NSData! { get }
    }

    /**
    Builds legal `multipart/form-data` parts, conforming to MultipartProtocol
    */
    class Multipart {
        
        class func json(name:String, parameters:[String:AnyObject]) -> MultipartProtocol? {
            return MultipartJson(name:name, parameters: parameters) }
        
        class func plain(name:String, text:String) -> MultipartProtocol? {
            return MultipartPlain(name:name, text:text) }
        
        /**
        Constructs an `application/json` part in a `multipart/form-data` body
        */
        class MultipartJson : MultipartProtocol {
            
            init?(name:String, parameters:[String:AnyObject]) {
                
                // initialize header
                self.headers = MultipartHeader()
                    .contentType(.JSON)
                    .contentDisposition(.Name(name))
                
                // serialize body
                let options = NSJSONWritingOptions.allZeros
                var error:NSError?
                self.body = NSJSONSerialization.dataWithJSONObject(parameters, options: options, error: &error)
                
                if self.body == nil {
                    return nil
                }
            }
            
            /// The headers used to generate an `application/json` part in a `multipart/form-data` part
            let headers:MultipartHeader
            
            /// The body used to generate an `application/json` part in a `multipart/form-data` part
            var body:NSData!
        }
        
        /**
        Constructs a `text/plain` part in a `multipart/form-data` body
        */
        class MultipartPlain : MultipartProtocol {
            
            init?(name:String, text:String) {
                
                // initialize header
                self.headers = MultipartHeader()
                    .contentType(.Plain)
                    .contentDisposition(.Name(name))
                
                // serialize body
                self.body = text.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false)
                if self.body == nil {
                    return nil
                }
            }
            
            /// The headers used to generate a `text/plain` part in a `multipart/form-data` part
            let headers:MultipartHeader
            
            /// The body used to generate a `text/plain` part in a `multipart/form-data` part
            var body:NSData!
        }
    }

    /**
    Defines legal `MIME` type values for the `Content-Type` header of `multipart/form-data`
    */
    enum Mime : String {
        case MultipartMixed = "multipart/mixed"
        case JSON = "application/json"
        case Plain = "text/plain"
    }

    /**
    Defines legal Disposition values for the `Content-Disposition` header of `multipart/form-data`
    */
    enum Disposition {
        case FormData
        case Name(String)
        case FileName(String)
        
        var encoded:String {
            switch self {
                
            case FormData:
                return "form-data"
                
            case Name(let name):
                return "name=\"\(name)\""
                
            case FileName(let name):
                return "filename=\"\(name)\""
            }
        }
    }

    /**
    Constructs `multipart/form-data` headers via the method chaining pattern.
    */
    class MultipartHeader {
        
        enum Key : String {
            case ContentType = "Content-Type"
            case ContentDisposition = "Content-Disposition"
        }
        
        /**
        Defines the `Content-Type` header and returns the instance of MultipartHeader
        upon which this call is made
        
        :param: mimeType  The desired MIME type, as defined by the Mime enumeration
        
        :returns: The instance of Multipart header upon which this call is made
        */
        func contentType(mimeType:Mime) -> MultipartHeader {
            headers[Key.ContentType.rawValue] = "\(mimeType.rawValue)"
            return self
        }
        
        
        /**
        Defines the `Content-Disposition` header with a single disposition value and
        returns the instance of MultipartHeader upon which this call is made
        
        :param: mimeType  The desired MIME type, as defined by the Disposition enumeration
        
        :returns: The instance of Multipart header upon which this call is made
        */
        func contentDisposition(disposition:Disposition) -> MultipartHeader {
            
            headers[Key.ContentDisposition.rawValue] =
                Disposition.FormData.encoded + "; " + disposition.encoded
            return self
        }
        
        /**
        Defines the Content-Disposition header with multiple disposition value and
        returns the instance of MultipartHeader upon which this call is made
        
        :param: mimeType  The desired MIME type, as defined by the Disposition enumeration
        
        :returns: The instance of Multipart header upon which this call is made
        */
        func contentDisposition(values:Disposition...) -> MultipartHeader {
            
            headers[Key.ContentDisposition.rawValue] = Disposition.FormData.encoded + "; "
                + "; ".join(map(values) { (d:Disposition) in d.encoded})
            return self
        }
        
        /// The resulting set of `multipart/form-data` headers
        var headers:[String:String] = [Key.ContentDisposition.rawValue: Disposition.FormData.encoded]
    }

    extension MultipartHeader : Printable {
        var description:String {
            return "Headers: \n\t"
                + "\n\t".join(map(self.headers.keys) { key in "\(key): \(self.headers[key]!)" })
        }
    }

**Summary:**

With the modest addition of the single function override above, I was able to add support for additional data types.  I hope that you can merge this change with your repository.  I also think that the extensions above for supporting additional types are clean, type-safe, and error resilient. I hope that it could help provide a path to extended type support.

**Comment by cnoon** [comment]
Hi @chrisco314, first off, my apologies for not replying to your Stack Overflow question. I've been focusing mainly on the 1.3.0 release and haven't been able to spend as much time on Stack Overflow as I would like. 

With that said, I'm fairly certain you should be able to do this already without any changes to Alamofire. Would it be possible for you to test a theory for me? Could you try to upload some JSON data to your server using the following approach?

```swift
multipartFormData.appendBodyPart(
    fileData: data, 
    name: "name", 
    fileName: "filename.json", 
    mimeType: "application/json"
)
```

I'm fairly certain this method should do "exactly" what it is you are looking for. I'm also curious if your variation is the exact same call minus the `filename` argument in the `Content-Disposition` header. We've had a similar issue opened in #569 but we never heard back from @top2tech. 

If you could verify whether that method works for your case that would be great. Additionally, it would also be useful to know if you are finding that the `filename` argument is optional. I'm having a difficult time determining from the RFC specs whether the filename is optional or not.

**Comment by chrisco314** [comment]
Short answer: No, I needed `text/plain` in the end

**In more detail:**

In the end, I needed to send plain text and not JSON (I did not have complete information when I started).  That meant that I had to manually create a "Multipart" object for `text/plain`, hence the method.  That method though, is as simple as it gets, is orthogonal, and could / would generally be the base two which others defer.

**Approach taken** 

To achieve my final goal, I could have hacked a couple more one-off functions and been done, but the approach I took allowed me a typesafe and modular means of building up Multipart objects, and by chaining the headers - themselves comprised of any combination of values.  For example, I noticed two function overloads for Content-Disposition, but neither worked for me and it didn't seem right to keep writing new combinations.  To that end, I added the variadic overload that would allow the flexibility to add any combination of fields:

        func contentDisposition(values:Disposition...) -> MultipartHeader {
            
            headers[Key.ContentDisposition.rawValue] = Disposition.FormData.encoded + "; "
                + "; ".join(map(values) { (d:Disposition) in d.encoded})
            return self
        }

**Improvements to the Api** 

While I am on the topic, I think the API for this functionality could be improved.  The `multipartFormData` callback feels neither necessary nor complete.  looking at the code, it only really just creates a generic `MultipartFormData` object and passes it back.  This could be done much more easily by having a default parameter on upload that constructed one.  The problem with the current form is that it turns out that calls to `appendBodyPart` can fail, and there is no way to stop the train at that point, you just get missing data on the server.

I am newer to functional programming, and I appreciate that you have kept your API dogma free, but going further (and cleaner), you could instead offer a failable initializers on `MultipartFormData` or method chaining that returns an optional self, to arrive at the following (more modular) approach:

    Alamofire.upload(
        authorizedRequest(.Attachments, method: .POST,  authToken: authToken),
        multipartFormData: MultipartFormData(
            Multipart.plain("id", value:last.id),
            Multipart.json("someJsonData", details:details),
            Multipart.file(firstFile, name: "file1"),
            Multipart.file(secondFile, name: "file2")),
        encodingCompletion: { encodingResult in
            // ..
        })
    
or
    
    Alamofire.upload(
        authorizedRequest(.Attachments, method: .POST,  authToken: authToken),
        multipartFormData: MultipartFormData()
            .appendBodyPart(Multipart.plain("id", value:last.id!))
            ?.appendBodyPart(Multipart.json("someJsonData", parameters:parameters))
            ?.appendBodyPart(fileURL: fileToUpload, name: "file1") {
            ?.appendBodyPart(fileURL: someOtherFile, name: "file2") {
        encodingCompletion: { encodingResult in
            // ..
        })

In the first case , you would use a variadic initializer taking optional values, and in the second use method chaining that returns optionals, to produce a final optional value that could then be accepted or rejected by `upload` depending on whether or not the optional `MultipartFormData` was valid.  To me, this is simpler, cleaner, as it is easy to construct, and fails immediately and decisively.  Both forms are equivalent, but I find the former to be simpler-looking, and more concise.

The use of optionals could also facilitate the functional / monad style of chaining, though the API would by no means be impacted by this, or carry any baggage to support it - it's just a potential bonus.  (I am by no means an expert in this area, I've just seen it used, and understand optionals to be the foundation for it)

**Comment by cnoon** [comment]
Okay @chrisco314, let's take a step back for a sec.

I need you to help me diagnose the problem you are having before I'm going to consider redesigning the APIs. Could you please answer my original questions? Can you directly pass in your `mimeType` to the following method and still have it work? If not, could you please walk me through why?

```
multipartFormData.appendBodyPart(
    fileData: data, 
    name: "name", 
    fileName: "filename.json", 
    mimeType: "application/json"
)
```

**Comment by chrisco314** [comment]
You do have a point, my issue when looking through that code initially though, was that it suggested a file, assumed a file, and hardcoded `filename` into the header.  In hindsight, I wish I had tested that, but thought at the time that it would cause problems with the server and felt that needed more control over the headers (for full disclosure, I have worked a lot with networking API's, but I was testing testing and inspecting headers with `curl` and `nc -L`).  To the point though, note that the function you suggest defers to this:

    private func contentHeaders(#name: String, fileName: String, mimeType: String) -> [String: String] {
        return [
            "Content-Disposition": "form-data; name=\"\(name)\"; filename=\"\(fileName)\"",
            "Content-Type": "\(mimeType)"
        ]
    }

To be fair, in my case, after I realized (in at least one case) that the server would not take json, I could have used this function that is implicitly hardcoded to assume `text/plain`, by virtue of it not taking a MIME type (which could then have just as easily taken data in String form):

    public func appendBodyPart(#data: NSData, name: String) {
        let headers = contentHeaders(name: name)
        let bodyStream = NSInputStream(data: data)
        let bodyContentLength = UInt64(data.length)
        let bodyPart = BodyPart(headers: headers, bodyStream: bodyStream, bodyContentLength: bodyContentLength)

        self.bodyParts.append(bodyPart)
    }

I think at that point, I already had my code written, and felt that I would need JSON later on.

I do think though that my original point remains with the proposed function though.  Since it assumes nothing at all - it can serve all needs.  Far from being a special case, it would be the absolute most general case, the function through which all related code-paths could flow.  

In summary though, my the key point here is the function that you suggest is assuming too much, and in doing so, is less flexible and extensible, and can consequently the API can be more difficult and less intuitive to apply to a wider range of cases where servers need particular header attributes.  Instead of trying to anticipate future needs, you could instead make it simpler, more general, and adopt any good adaptations that come out of it as a bonus.

As a bit of an side, I still feel that my other points remain, that the means of configuring the `MultipartFormData` could be simpler while again being more general, and consequently more flexible. Frankly speaking, when using the multipart-form API, I was left with the impression that it deviated from the modular-construction form of the rest of the API - .GET / .POST etc. calls, that allowed you to build your own request object, and configure your headers.  To me, it was different enough that it felt like two libraries had been merged - it seemed to break the existing conventions.  I do think the approach of failable initializers presents a simpler, more intuitive alternative, and would not be hard to implement.

I do appreciate your hard work though; I don't mean this to sound like a criticism.  It can't be easy working through all the cases.  

Thanks

**Comment by cnoon** [comment]
Okay, so if I boil down your extremely verbose set of multiple answers, you are able to do what you need with either of the following methods:

* `appendBodyPart(fileData:name:fileName:mimeType:)`
  > If you provide a dummy `fileName` value

* `appendBodyPart(data:name:)`
  > If you override it and add a `mimeType` argument that would append a `Content-Type` header

As an initial takeaway, I'd say the `appendBodyPart(data:name:)` method most likely needs an optional `mimeType` parameter which will add the `Content-Type` header to the body part if it is passed in. That would certainly solve your main issue.

---

As for your proposed API changes, I really appreciate your feedback! 

When designing the current implementation, the one thing that didn't feel exactly right was the fact that you can easily bury the errors that occur when appending a new body part if you aren't careful. Only some of the methods can actually fail, but if they fail without you handling them, your body part won't make it up to the server.

As for providing the closure in the `upload` method, that is simply to be able to create the data inline with the rest of the request. I'll consider adding another alternative API where you pass the `MultipartFormData` object in directly rather building it on the fly with a provided closure.

Overall, you have piqued my interest here so I will probably be making some changes to address some if not all of these issues. Once I push up a PR I'd appreciate some more feedback if you have time.

Cheers 🍻

**Comment by chrisco314** [comment]
Right, pretty close.  Apologies for the verbosity, I was trying to be unambiguous without being curt, here is the digest version:

On the topic of `appendBodyPart` (and for all code in general), don't assume more than you have to:

1. `appendBodyPart(fileData:name:fileName:mimeType:)` is unnecessarily specific in that it assumes a file
1. `appendBodyPart(data:name:)` is unnecessarily restrictive, it assumes a plain text payload
1. `appendBodyPart(data:name:type:)` would be unnecessarily restrictive as it would restrict other headers like `content-transfer-encoding`
1. don't try to anticipate all options, present and future, allow at least one "designated initializer" that assumes nothing and allows the caller to specify everything they need, make this the base call through which all others flow
1. and nitpicky, but general convention is that key comes before value, header comes before body, so the order of the parameters `data`, `name` seems unnatural

On the topic of `upload`:

1. don't add a callback unless it adds value (does asynchronous processing), in this case, it merely adds complexity and decreases flexibility
1. there are no obvious ways to do error handling when adding parts to `MultipartFormData`, the closure is unnecessary and prevents error handling, forcing an upload even when an error is known.  You can detect an error, but the train has left the station and you cannot stop it since the callback has no return value
1. keep things as simple as possible, user constructed instance type parameters are simpler than closures, `upload` overrides could provide pre-configured `MultipartFormData` instances
1. failable initializers offer a graceful way of chaining operations, while catching results, and without any baggage, allow more modern programming techniques - see Monads and "railway oriented programming"

And having since worked with `download`:

* it takes unnecessary effort to get the final download name, see multiple posts (not of my own) on Stackoverflow, here is one example: http://stackoverflow.com/questions/26307170/alamofire-download-method-where-is-the-file-and-did-it-save-successfully

Cheers

**Comment by cnoon** [comment]
Okay @chrisco314, I just pushed up PR #596 which addresses all the issues presented in this ticket. Please check it out and redirect all further comments to #596.

### Proposals Implemented

I took about half of your suggestions in this ticket and implemented a variation of them. 

* Added an `appendBodyPart(data:name:mimeType:)` method which solves your root issue. 
* Added a base method that everything calls through that is completely wide open allowing for headers like `Content-Transfer-Encoding`. 
* Now internally tracking a body part error that is bubbled up immediately during encoding. This eliminates the possibility of uploading something to the server that is missing a body part.

### Proposals Not Implemented

Some of the suggestions that I did not incorporate were:

* Reordering the parameter names

  > The parameter names begin with either `data`, `fileURL` or `stream` to make it easier to work with autocomplete to narrow down your choices. Starting with `name` makes it difficult to pick out the one you are looking for. While I completely understand your logic, I disagree that this would be the best course of action.

* Variadic Initializer and/or Chaining Optionals

> Both of these approaches make the APIs clunky and more difficult to use IMO.

* Multiple Enumerations and Chained Headers

  > This approach presents way too many ways for users to make mistakes. It also does not provide any flexibility since you cannot customize the behavior by extending enumeration cases.

### Summary

Alamofire is intended to provide very clean, robust APIs to make networking easy, yet be flexible enough to allow users to customize any part of the networking stack, thus supporting a range of experience from novice to expert. Therefore, some of your suggestions aligned very well with our overalls goals which is exactly why they were implemented.

Thanks again for all your efforts here...Alamofire is certainly better for it!

Cheers 🍻

**Comment by pavm035** [comment]
Hi 

How to support multipart/mixed as content type?

Thanks

**Comment by cnoon** [comment]
Hi @pavm035, multipart/mixed uploads are not supported at this time.
